cpp/include/cudf/dictionary/dictionary_factories.hpp

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/null_mask.hpp>
+
+
+namespace cudf
+{
+
+/**
+ * @brief Construct a dictionary column by dictionary encoding an existing column.
+ *
+ * Unique elements are managed in a dictionary.
+ * The output column will have indices with the same count
+ * as the input column.
+ *
+ * ```
+ * c = [429,111,213,111,213,429,213]
+ * d = make_dictionary_column(c)
+ * d now has keys [111,213,429] and indices [2,0,1,0,1,2,1]
+ * ```
+ *
+ * @param[in] column The column to dictionary encode.
+ * @param[in] mr Optional resource to use for device memory allocation.
+ * @param[in] stream Optional stream on which to issue all memory allocation and
+ * device kernels.
+ */
+std::unique_ptr<column> make_dictionary_column(
```

Comment:
I don't mind creating an `encode` function too but I think people will likely be expecting to find a function called `make_dictionary_column` based on the pattern we have created for factory functions. Maybe one can just call the other.

Reply:
Most (all?) other factories don't take a `column/_view` as input and do work on it to create a new column.

For example, the strings column factory accepts a `column_view` for the chars/indices which it just copies into a new `column`. 

Likewise, the `make_dictionary_column` that accepts `column_view`s for the keys/indices is appropriately a "factory" function.

I think `encode` is different from a factory because isn't just copying the input `column_view`, it's performing work on the `input` column to create new ones.

I think factories should be limited to cases where we're either just allocating memory or copying existing columns. Any cases where we're doing additional work make more sense to me to be independent, descriptive functions like `encode`. 

That said, having a factory that calls the `encode` is fine compromise in my mind.

Reply:
Ok, I like this explanation. I agree this `encode` logic should not be a factory method.
Perhaps this description should go into the Transition Guide or some similar document.